KFSPTS-25263 create base batch job to upload suppliers to Jaggaer

jhulslander opened the pull request:
Please note there is a related non-prod SQL PR.  Don't merge this till both PRs are ready.
There will be a follow up user story to actually create the CSV file, and to make use of the FIPS to ISO mapping service.

I noticed JaggaerUploadDaoJdbc does include some static final variable that I think exist in base code.  I think this PR is ready for review, but please note I want to look into adjusting those variables

ajd299 reviewed (changes requested):
Please see my suggested changes, most of them are just typos

jhulslander commented:
retest this please

jhulslander commented on src/main/java/edu/cornell/kfs/module/purap/batch/JaggaerGenerateContractPartyCsvStep.java:
```diff
@@ -0,0 +1,86 @@
+        return getParameterValueString(CUPurapParameterConstants.JAGGAER_UPLOAD_VENDOR_DATE);
+    }
+    
+    protected String getParameterValueString(String paramterName) {
+        return parameterService.getParameterValueAsString(this.getClass(), paramterName);
+    }
```
ok

jhulslander commented on src/main/java/edu/cornell/kfs/module/purap/batch/dataaccess/impl/JaggaerUploadDaoJdbc.java:
```diff
@@ -0,0 +1,160 @@
+            };
+            return this.getJdbcTemplate().query(buildVendorSql(processingMode, processingDate), rowMapper);
+        } catch (Exception e) {
+            LOG.error("findJaggaerContractParty, had an error finding jaggaer contract partiess: ", e);
```
ok

jhulslander commented on src/main/java/edu/cornell/kfs/module/purap/batch/dataaccess/impl/JaggaerUploadDaoJdbc.java:
```diff
@@ -0,0 +1,160 @@
+        return sql;
+    }
+    
+    private String buildPurchasOrderLimitSubQuery(String processingDate) {
```
ok

jhulslander commented on src/main/java/edu/cornell/kfs/module/purap/batch/dataaccess/impl/JaggaerUploadDaoJdbc.java:
```diff
@@ -0,0 +1,160 @@
+        return sb.toString();
+    }
+    
+    private String buildVendorNumber(String vendorHeaderId, String venderDetailId) {
+        return vendorHeaderId + KFSConstants.DASH + venderDetailId;
```
ok

jhulslander commented on src/main/java/edu/cornell/kfs/module/purap/batch/service/impl/JaggaerGenerateContractPartyCsvServiceImpl.java:
```diff
@@ -0,0 +1,65 @@
+        return uploadDtos;
+    }
+    
+    protected List<JaggaerContractAddressUploadDto> findVendorAddresses(List<JaggaerContractAddressUploadDto> allAddresses, String erpNymber) {
+        return allAddresses.stream()
+                .filter(address -> StringUtils.equalsIgnoreCase(address.getERPNumber(), erpNymber))
+                .collect(Collectors.toCollection(ArrayList::new));
+    }
```
ok

jhulslander commented on src/main/java/edu/cornell/kfs/module/purap/businessobject/lookup/JaggaerContractPartyUploadDto.java:
```diff
@@ -0,0 +1,141 @@
+    public String toString() {
+        ReflectionToStringBuilder builder = new ReflectionToStringBuilder(this, ToStringStyle.NO_CLASS_NAME_STYLE);
+        builder.setExcludeFieldNames("taxIdentificationNumber");
+        builder.append("taxIdentificationNumber", StringUtils.isBlank(taxIdentificationNumber) ? StringUtils.EMPTY : "restrctived tax id number");
```
switched to Restricted

jhulslander commented on src/main/java/edu/cornell/kfs/vnd/CUVendorConstants.java:
```diff
@@ -19,6 +19,8 @@
 import java.util.LinkedHashMap;
 import java.util.Map;
 
+import edu.cornell.kfs.vnd.CUVendorConstants.FIELD_NAMES;
+
```
nope, not needed, thanks

jhulslander commented on src/main/java/edu/cornell/kfs/vnd/CUVendorConstants.java:
```diff
@@ -79,4 +81,23 @@ public static class CUAddressTypes {
     
     public static final int MAX_VENDOR_CONTACT_NAME_LENGTH = 45;
+    
+    public final class FIELD_NAMES {
```
yeah, it should be static

jhulslander commented on src/test/java/edu/cornell/kfs/module/purap/batch/service/impl/JaggaerGenerateContractPartyCsvServiceImplTest.java:
```diff
@@ -0,0 +1,191 @@
+        assertEquals(4, jaggaerUploadDtos.size());
+        
+        JaggaerContractPartyUploadDto actualFirstELement = (JaggaerContractPartyUploadDto) jaggaerUploadDtos.get(0);
+        assertEquals(buildVendor1Dto(), actualFirstELement, "The first element shound be vendor 1");
```
ok

jhulslander commented on src/test/java/edu/cornell/kfs/module/purap/batch/service/impl/JaggaerGenerateContractPartyCsvServiceImplTest.java:
```diff
@@ -0,0 +1,191 @@
+        expectedAddress2.setName(actualFirstELement.getContractPartyName());
+        assertEquals(expectedAddress2, actualThirdElement, "The third element should be vendor 1 address 2");
+        
+        JaggaerContractPartyUploadDto actualFourthELement = (JaggaerContractPartyUploadDto) jaggaerUploadDtos.get(3);
```
ok

jhulslander commented on src/test/java/edu/cornell/kfs/module/purap/batch/service/impl/JaggaerGenerateContractPartyCsvServiceImplTest.java:
```diff
@@ -0,0 +1,191 @@
+        assertEquals(expectedAddress2, actualThirdElement, "The third element should be vendor 1 address 2");
+        
+        JaggaerContractPartyUploadDto actualFourthELement = (JaggaerContractPartyUploadDto) jaggaerUploadDtos.get(3);
+        assertEquals(buildVendor2WithTaxIdDto(), actualFourthELement, "THe fourth elemeent should be vendor 2");
```
ok

jhulslander commented on src/test/java/edu/cornell/kfs/module/purap/batch/service/impl/JaggaerGenerateContractPartyCsvServiceImplTest.java:
```diff
@@ -0,0 +1,191 @@
+    
+    static Stream<Arguments> jaggaerContractPartyUploadDtoToStringParameters() {
+        return Stream.of(
+                Arguments.of(buildVendor2WithTaxIdDto(), "taxIdentificationNumber=restrctived tax id number"),
```
restricted

jhulslander commented:
I've made the changes suggested, I also discovered a problem with the SQL that I have corrected, in doing so I simplified and reduced some duplication in how the SQL is generated in JaggaerUploadDaoJdbc.

This is ready for review again

ajd299 reviewed (approved):
Code looks good, merging
